### PR TITLE
[LC-16] apply websocket for citizen to mother peer connection

### PR DESCRIPTION
**Pull request by laurenceyoon**
## apply websocket for citizen to mother peer connection

* How it works
  * The websocket connection replaces the existing `node_Subscribe` and `node_AnnounceConfirmedBlock`.
  * After the block height sync of the Citizen node is over, Citizen sends a subscribe request to the websocket, bypassing its current block height. (loopchain side)
  * This request is received from the icon-rpc-server dispatcher of the mother peer, and it passes the block data whenever a new block is generated. (iconrpcserver side)
* Delete announce_confirmed_block(broadcast_scheduler) logic.
* For compatibility, the existing announce_confirmed_block in channel_inner_service remain for some time.

### CODE REVIEW

File: loopchain/channel/channel_inner_service.py [hunk 1]
```diff
@@ -38,11 +40,30 @@ class ChannelInnerTask:
...
         self._channel_service = channel_service
         self._thread_pool = ThreadPoolExecutor(1, "ChannelInnerThread")
+        self._condition_new_block: Condition = None
 
     @message_queue_task
     async def hello(self):
         return 'channel_hello'
 
+    @message_queue_task
+    async def subscribe_new_block(self, citizen_block_height: int):
```
Comment (windies21) [inline loopchain/channel/channel_inner_service.py]: It need subscription limitation.

File: loopchain/channel/channel_inner_service.py [hunk 2]
```diff
@@ -38,11 +40,30 @@ class ChannelInnerTask:
...
+
+        if citizen_block_height == peer_block_height:
+            async with self._condition_new_block:
+                await self._condition_new_block.wait()
+
+        block_manager = self._channel_service.block_manager
+        new_block_height = citizen_block_height + 1
+        new_block = block_manager.get_blockchain().find_block_by_height(new_block_height)
+        if new_block is None:
+            fail_response_code = message_code.Response.fail_wrong_block_height
```
Comment (zsaladin) [inline loopchain/channel/channel_inner_service.py]: This code is not affected. I think it has to raise exception.

File: loopchain/channel/channel_inner_service.py [hunk 3]
```diff
@@ -593,9 +614,21 @@ def stop(self, message):
...
+        self._task._condition_new_block = Condition(loop=self.loop)
+
     def _callback_connection_lost_callback(self, connection: RobustConnection):
         util.exit_and_msg("MQ Connection lost.")
 
+    def notify_new_block(self):
+        async def _notify():
+            condition = self._task._condition_new_block
+            async with condition:
+                condition.notify()
```
Comment (zsaladin) [inline loopchain/channel/channel_inner_service.py]: notify() -> notify_all()

File: loopchain/baseservice/websocket_client.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+                    await self.__stop_shutdown_timer()
+                    request = json.dumps({
+                        'height': self.__block_height
+                    })
+                    await websocket.send(request)
+
+                    while True:
+                        new_block_json = await websocket.recv()
+                        response_code = await self.__add_confirmed_block(block_json=new_block_json)
+                        if response_code == message_code.Response.success:
```
Comment (zsaladin) [inline loopchain/baseservice/websocket_client.py]: This 'if' statement can be removed, if `self.__add_confimed_block` raise exception.

File: loopchain/channel/channel_inner_service.py (hunk 1 above)
Comment (laurenceyoon) [reply loopchain/channel/channel_inner_service.py]: subscription limit has been added.

File: loopchain/channel/channel_inner_service.py [hunk 5]
```diff
@@ -38,11 +40,30 @@ class ChannelInnerTask:
...
 
     @message_queue_task
     async def hello(self):
         return 'channel_hello'
 
+    @message_queue_task
+    async def subscribe_new_block(self, citizen_block_height: int):
+        peer_block_height = self._channel_service.block_manager.get_blockchain().block_height
+        if citizen_block_height > peer_block_height:
+            raise RuntimeError
```
[the remark this answers is not in the document]
Comment (laurenceyoon) [reply loopchain/channel/channel_inner_service.py]: Updated: when citizen's block height is higher than peer's, the mother peer returns error message.

File: loopchain/channel/channel_inner_service.py (hunk 3 above)
Comment (laurenceyoon) [reply loopchain/channel/channel_inner_service.py]: Completed.

File: loopchain/baseservice/websocket_client.py (hunk 4 above)
Comment (laurenceyoon) [reply loopchain/baseservice/websocket_client.py]: Removed.

**Review by windies21 — APPROVED** [review]

**Pull request merged**